src/g1.rs

```diff
@@ -399,15 +399,21 @@ impl G1Affine {
     /// exists within the $q$-order subgroup $\mathbb{G}_1$. This should always return true
     /// unless an "unchecked" API was used.
     pub fn is_torsion_free(&self) -> Choice {
-        const FQ_MODULUS_BYTES: [u8; 32] = [
-            1, 0, 0, 0, 255, 255, 255, 255, 254, 91, 254, 255, 2, 164, 189, 83, 5, 216, 161, 9, 8,
-            216, 57, 51, 72, 125, 157, 41, 83, 167, 237, 115,
-        ];
+        // Algorithm from Section 6 of https://eprint.iacr.org/2021/1130.
+        //
+        // Check that endomorphism_p(P) == -[X^2]P
+
+        // An early-out optimization described in Section 6.
+        // If uP == P but P != point of infinity, then the point is not in the right
+        // subgroup.
+        let x_times_p = G1Projective::from(self).mul_by_x().neg();
+        if G1Affine::from(x_times_p).eq(self) && !bool::from(self.infinity) {
+            return Choice::from(0u8);
+        }
 
```

Comment:
This does not involve secret information. Does it make sense to consider constant-time for this algorithm?

Reply:
We do not know at this layer whether `self` is a secret or not; that depends on the protocol the G1 point is being used in.

The general approach we've taken in this crate (and one of the primary differences from the previous BLS12-381 implementation in `pairing`) is to use constant time code as much as possible, so that we provide a secure base dependency for others to build on. See also e.g. our constant-time implementation of Tonelli-Shanks for square roots.

In particular, it's easy to use constant time dependencies inside non-constant-time logic, but it's impossible to do the reverse.

Reply:
Ah, sorry! I didn't note the API guarantees constant time (although the `Choice` should have been a hint)